presto-main/src/main/java/com/facebook/presto/type/TimestampOperators.java

```diff
@@ -159,10 +194,21 @@ public static Slice castToSlice(ConnectorSession session, @SqlType(StandardTypes
     public static long castFromSlice(ConnectorSession session, @SqlType("varchar(x)") Slice value)
     {
         try {
-            return parseTimestampWithoutTimeZone(session.getTimeZoneKey(), trim(value).toStringUtf8());
+            if (session.isLegacyTimestamp()) {
+                return parseTimestampWithoutTimeZone(session.getTimeZoneKey(), trim(value).toStringUtf8());
+            }
+            else {
+                return parseTimestampWithoutTimeZone(trim(value).toStringUtf8());
+            }
         }
         catch (IllegalArgumentException e) {
-            throw new PrestoException(INVALID_CAST_ARGUMENT, "Value cannot be cast to timestamp: " + value.toStringUtf8(), e);
+            try {
+                parseTimestampWithTimeZone(session.getTimeZoneKey(), trim(value).toStringUtf8());
+                throw new PrestoException(INVALID_CAST_ARGUMENT, "Value cannot be cast to timestamp as it contains explicitly defined TIME ZONE: " + value.toStringUtf8(), e);
```

Comment:
This behavior is incorrect.

There was a previous discussion about how this should behave in #3098.

According to SQL Spec 2016 Foundation 6.13 General Rules 18 a ii, 

> 1) If the rules for `<literal>` or for `<unquoted timestamp string>` in Subclause 5.3, “`<literal>`”, can be applied to SV to determine a valid value of the data type TD, then let TV be that value.
> 
> 2) If the rules for `<literal>` or for `<unquoted timestamp string>` in Subclause 5.3, “`<literal>`”, can be applied to SV to determine a valid value of the data type TIMESTAMP(TSP) WITH TIME ZONE, then let TV be the value of:
> 
> ```
> CAST ( CAST ( VE AS TIMESTAMP(TSP) WITH TIME ZONE ) AS TIMESTAMP(TSP) WITHOUT TIME ZONE )
> ```
>
> 3) Otherwise, an exception condition is raised: data exception — invalid datetime format.

Therefore, the correct behavior is to parse the string, and ignore the whatever the string says about time zone. 

Note that the conclusion here is different from what we reached in #3098 (because CAST from TIMESTAMP WITH TIME ZONE to TIMESTAMP WITHOUT TIME ZONE had a different semantics).

Reply:
Good catch